### PR TITLE
Turn off LangChain autolog artifact logging in favor of the tracer

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/11634?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11634/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11634
```

</p>
</details>

### What changes are proposed in this pull request?

The current production LangChain autologging does followings:
1. Logging intermediate steps like LLM input/output using [MlflowLangchainCallbackHandler](https://github.com/langchain-ai/langchain/blob/38fb1429fe5a955a863fb91b626c2c9f85efb703/libs/community/langchain_community/callbacks/mlflow_callback.py#L256).
2. Logging final artifacts such as model, signature, input example, etc, at the completion of the prediction. These are optional and only logged when corresponding configuration is enabled e.g. `mlflow.langchain.autolog(log_model_signatures=True)`.

However, (1) largely overlaps with the traces. Most of metrics/jsons logged at each intermediate steps are subset of the information available in traces. The duplication itself is not the main issue, but its side effect is. The legacy callback logs metrics/json at each step synchronously, so it adds certain latency (~1 sec) to the each step execution. This latency is not captured by the tracer because callbacks are executed independently, results in weird blanks in the span timeline.

Therefore, we decided to turn off the legacy callback in the feature branch. This PR removes (1) while keep (2) to be functioning, with some refactoring.

**Note**: There is one feature in the legacy callback that is covered by the new tracer - [complex text metrics](https://github.com/langchain-ai/langchain/blob/38fb1429fe5a955a863fb91b626c2c9f85efb703/libs/community/langchain_community/callbacks/mlflow_callback.py#L62-L81). These metrics are computed against LLM output inside the legacy callback, but we don't have corresponding implementation in the tracer. This should not be a blocker for feature branch, but we should add this back before the public release, probably as a separate callback.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/langchain/_langchain_autolog.py
Comment (serena-ruan) [inline mlflow/langchain/_langchain_autolog.py:115]: This is not correct as you created a run if there's no active run, then the current mlflow.active_run().info.run_id must be run_id.

File: mlflow/langchain/_langchain_autolog.py
Comment (serena-ruan) [inline mlflow/langchain/_langchain_autolog.py:151]: Why do we handle the run creation here instead of within the callback?

File: mlflow/langchain/_langchain_autolog.py
Comment (serena-ruan) [inline mlflow/langchain/_langchain_autolog.py:92]: Why do we need this? MAX_REQ_VERSION is dynamically determined, if the latest version works then we should update MAX_REQ_VERSION.

File: mlflow/langchain/_langchain_autolog.py
Comment (serena-ruan) [inline mlflow/langchain/_langchain_autolog.py:127]: This should be called after `result = original(self, *args, **kwargs)`

File: mlflow/langchain/__init__.py
Comment (serena-ruan) [inline mlflow/langchain/__init__.py:865]: Let's revert this, this default value has been discussed and we want to make it default to True to record user inputs & outputs, and it will serve as table for evaluation.

**Review by serena-ruan — COMMENTED** [review]
It seems we're just removing the original MLflowCallback. This sounds like a behavior change to the customers, while I agree most of the information are duplicated by traces. I think we can combine the them into the same callback as I did before https://github.com/serena-ruan/mlflow/blob/18e3e5abd1a7b4fa87dbb979598406c216278475/mlflow/langchain/_langchain_autolog.py#L251 and add param to turn on/off the metrics/artifacts logging. We need to add them back after all.

**Comment by B-Step62** [comment]
> It seems we're just removing the original MLflowCallback. This sounds like a behavior change to the customers, while I agree most of the information are duplicated by traces. I think we can combine the them into the same callback as I did before https://github.com/serena-ruan/mlflow/blob/18e3e5abd1a7b4fa87dbb979598406c216278475/mlflow/langchain/_langchain_autolog.py#L251 and add param to turn on/off the metrics/artifacts logging. We need to add them back after all.

I don't think it's good timing to edit the tracer considering the timeline for the preview, we can do that as follow-up. Even when we adding back, I'd still keep them separate callback given that they serves different purpose and the tracer implementation is already complex enough, unless there is particular merit in merging them.

File: mlflow/langchain/__init__.py
Comment (B-Step62) [reply mlflow/langchain/__init__.py:865]: If we revert this the default tracing experience creates a new run, which doesn't align with the decision.

> We don't create new runs automatically

File: mlflow/langchain/_langchain_autolog.py
Comment (B-Step62) [reply mlflow/langchain/_langchain_autolog.py:127]: great catch, thanks!

File: mlflow/langchain/_langchain_autolog.py
Comment (B-Step62) [reply mlflow/langchain/_langchain_autolog.py:92]: I haven't tested the all autologging logic with the newer version, and particularly for model logging.

File: mlflow/langchain/_langchain_autolog.py
Comment (B-Step62) [reply mlflow/langchain/_langchain_autolog.py:115]: What if user specifies `run_id`? We don't create a new run if it's set, and I thought that is the intention of the original logic.

Comment (serena-ruan) [reply mlflow/langchain/_langchain_autolog.py:115]: Yes, so we shouldn't use mlflow.active_run() here, we should assign a variable active_run=mlflow.active_run() before the autologging run creation step, and then check active_run here

**Comment by serena-ruan** [comment]
> > It seems we're just removing the original MLflowCallback. This sounds like a behavior change to the customers, while I agree most of the information are duplicated by traces. I think we can combine the them into the same callback as I did before https://github.com/serena-ruan/mlflow/blob/18e3e5abd1a7b4fa87dbb979598406c216278475/mlflow/langchain/_langchain_autolog.py#L251 and add param to turn on/off the metrics/artifacts logging. We need to add them back after all.
> 
> I don't think it's good timing to edit the tracer considering the timeline for the preview, we can do that as follow-up. Even when we adding back, I'd still keep them separate callback given that they serves different purpose and the tracer implementation is already complex enough, unless there is particular merit in merging them.

Then we can't avoid the original issue you want to solve in this PR, if multiple callbacks are invoked then there will be gaps between spans. If preview is based on this branch, then we can simply remove the original callback. The run handling logic is refactored, it shouldn't impact tracer as currently it's not used, but I think we'll need bug bash to make sure it works for artifacts logging afterwards, unless we want to remove the metrics/artifacts logging.

**Comment by serena-ruan** [comment]
We can use `manage_run=True` when patch the function, so we don't need to handle the run creation.

File: mlflow/langchain/_langchain_autolog.py
Comment (B-Step62) [reply mlflow/langchain/_langchain_autolog.py:115]: I see, so we should keep the original active run before starting a new one like [this](https://github.com/mlflow/mlflow/blob/master/mlflow/langchain/_langchain_autolog.py#L218). That makes sense, thanks for the description!

**Comment by dbczumar** [comment]
> > Then we can't avoid the original issue you want to solve in this PR, if multiple callbacks are invoked then there will be gaps between spans.
> 
> The issue this PR tries to solve is we have those gaps **by default** for tracing experience. It's ok if customer turn on some additional callback operation, like we can't stop customers from injecting custom callback that also creates gaps.
> 
> > If preview is based on this branch, then we can simply remove the original callback.
> 
> Yeah we can do that as the simplest form, but are we ok with dropping logging models/signature/etc at all? They are not covered by traces, which is why I tried to keep them but remove table/metrics logging alone. I don't have an answer myself, what do you think @dbczumar @BenWilson2 ?

Custom callback injection is probably a common use case for LangChain. Our tracing should still work - is there a way to make that happen?

The version of autologging that we release to customers needs to have model logging. If it's easy to keep that around while we refactor / adjust the autologging implementation, then IMO we should keep it around :D

File: mlflow/langchain/_langchain_autolog.py
Comment (B-Step62) [reply mlflow/langchain/_langchain_autolog.py:151]: The main reason is simply we turning off callback but need run for logging models/signatures/etc. But also it feels a bit weird that the callback creates a run and the rest of autologging logic (that is not related to callback at all) in MLflow relies on it.

**Comment by B-Step62** [comment]
> Custom callback injection is probably a common use case for LangChain. Our tracing should still work - is there a way to make that happen?

Hmm this might be a bit tricky

> The version of autologging that we release to customers needs to have model logging. If it's easy to keep that around while we refactor / adjust the autologging implementation, then IMO we should keep it around :D

Got it. I think it's easy, except the one complication that we extract the run management logic from the original callback to MLflow side. Imo it's the right direction eventually, but we definitely need bug-bash before merging back to master as you mentioned. WDYT @serena-ruan ?

File: mlflow/langchain/_langchain_autolog.py
Comment (BenWilson2) [inline mlflow/langchain/_langchain_autolog.py:153]: let's use the consistent `name_utils._generate_random_name(integer_scale=2)` name generator instead of the random characters here

File: mlflow/langchain/_langchain_autolog.py
Comment (B-Step62) [reply mlflow/langchain/_langchain_autolog.py]: The functionality works with the latest version as long as I've tested, so bumped up the max version as suggested.

File: mlflow/langchain/_langchain_autolog.py
Comment (B-Step62) [inline mlflow/langchain/_langchain_autolog.py:226]: note: There was a bug in original autologging - it cannot add tags to an existing active run, because the `resolved_tags` here contains reserved tags like `mlflow.user`. Those tags are already logged in the run when it started, and raise an exception when we try to override. The refactored logic includes fix.

**Comment by B-Step62** [comment]
@serena-ruan @BenWilson2 @dbczumar I've updated the PR as per the comments and a few bugs identified from the bug-bash and my testing. I need to do more rigorous bug-bash before public release to ensure that there is no regression in current autologging UJ (expect what can be replaced with trace functionality), but this version should be more or less ready for the PrPr. Would appreciate if you guys could take another look, thanks!

File: mlflow/langchain/_langchain_autolog.py
Comment (B-Step62) [inline mlflow/langchain/_langchain_autolog.py:119]: note: updating to context manager to ensure termination of the run

File: mlflow/entities/span_event.py
Comment (BenWilson2) [inline mlflow/entities/span_event.py:51]: couldn't this also just be a static method of the class? The error before was because it was declared as an instance method right?

File: mlflow/langchain/_langchain_autolog.py
Comment (BenWilson2) [inline mlflow/langchain/_langchain_autolog.py:68]: Sorry for the silly nit here, but should this be a classmethod?

```suggestion
    @classmethod
    def init(cls):
        config_dict = AUTOLOGGING_INTEGRATIONS.get(mlflow.langchain.FLAVOR_NAME, {})
        return cls(
            log_models=config_dict.get("log_models", False),
            log_input_examples=config_dict.get("log_input_examples", False),
            log_model_signatures=config_dict.get("log_model_signatures", False),
            log_inputs_outputs=config_dict.get("log_inputs_outputs", False),
            extra_tags=config_dict.get("extra_tags", None),
        )
```

File: mlflow/langchain/_langchain_autolog.py
Comment (BenWilson2) [inline mlflow/langchain/_langchain_autolog.py:81]: When testing this it really makes we wish that we implemented a global state variable that kept track if this warning has been issued already in a session and to suppress all other warnings once the first one popped up.

File: mlflow/langchain/_langchain_autolog.py
Comment (BenWilson2) [inline mlflow/langchain/_langchain_autolog.py:150]: Let's be sure to test the random name generation here. With a very short max_length the probability of collisions is VERY high. See the implementation of the generator and how the names are constructed to be doubly certain that we aren't potentially creating collisions here with high run_id counts / experiment

File: mlflow/langchain/_langchain_autolog.py
Comment (BenWilson2) [inline mlflow/langchain/_langchain_autolog.py:164]: +1 to this approach :)

File: mlflow/langchain/_langchain_autolog.py
Comment (BenWilson2) [inline mlflow/langchain/_langchain_autolog.py:173]: Should any tag that starts with the protected 'namespace' `mlflow.` be considered immutable? We have similar precedent with reserving tags like this elsewhere in MLflow and I'm just curious what your thoughts are about expanding the denylist here to not overwrite tags that may be added in the future without having the context of this denylist that is reduced in scope

**Review by BenWilson2 — APPROVED** [review]
A few nits / questions, but otherwise this contains a lot of important fixes. Thanks Yuki!!

LGTM once we validate that run name collisions are a very low probability / adjust the chances of collision

File: mlflow/langchain/_langchain_autolog.py
Comment (B-Step62) [reply mlflow/langchain/_langchain_autolog.py:150]: Hmm what is the impact of collision? Run ID should be unique but run name can be arbitral string.

Comment (BenWilson2) [reply mlflow/langchain/_langchain_autolog.py:150]: It makes the UI unusable ;)

Comment (B-Step62) [reply mlflow/langchain/_langchain_autolog.py:150]: Ah I see so it's UX concern (thought it's backend thing). Do you know if we have any constraint in run name length? If not we can just append `langchain-` to the normal 20-chars random name. If it's not possible, maybe we can use shorter prefix like `lc-`?

File: mlflow/langchain/_langchain_autolog.py
Comment (B-Step62) [reply mlflow/langchain/_langchain_autolog.py:68]: makes sense!

File: mlflow/entities/span_event.py
Comment (B-Step62) [reply mlflow/entities/span_event.py:51]: Yeah that works too. I though it's kinda better to define outside as it does nothing with span event, but this method will likely only used within this class so putting it as static method makes more sense now.

File: mlflow/langchain/_langchain_autolog.py
Comment (B-Step62) [reply mlflow/langchain/_langchain_autolog.py:173]: > Should any tag that starts with the protected 'namespace' mlflow. be considered immutable?

Strictly to say, only a subset of reserved tags are immutable in the Databricks backend (defined as `userImmutableRunTags` in universe). That said, excluding `mlflow.` should not harm practically, as it seems none of the reserved tags need to be altered here in autologging. It sounds better than defining another allowlist in OSS side.

File: mlflow/langchain/_langchain_autolog.py
Comment (B-Step62) [reply mlflow/langchain/_langchain_autolog.py:81]: Great point, I thought `warnings` does manage such state, but it seems they only suppress warning if multiple same warnings are issued within a single display call. It doesn't deduplicate warning across multiple calls, so the warning will shown repeatedly across test cases, notebook cells, and etc.

File: mlflow/langchain/_langchain_autolog.py
Comment (B-Step62) [reply mlflow/langchain/_langchain_autolog.py:150]: Actually it's obvious that we shouldn't have strict length limit as we allow users to set run name, I'll just append the default 20 chars random name after `langchain-` prefix.

**Review by serena-ruan — APPROVED** [review]
LGTM! Thanks for the refactoring and run status fix!

File: mlflow/langchain/_langchain_autolog.py
Comment (B-Step62) [reply mlflow/langchain/_langchain_autolog.py:137]: The intention of defining them in each branch was to make the correspondence of branch and value clear (like switch statement). Let me keep it for now but can add default if we start to have more branches or each block becomes longer!

**Pull request merged**
